Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java

```diff
@@ -2359,44 +2402,110 @@ else if (startKind.equals("Item"))
 					throw new UserErrorMessage("You do not have physical access to this item so you cannot transfer anything to/from it. It needs to be near you or in your inventory.");
 				
 				// Check if the character can actually carry something else or if its all too heavy...
-				Long itemWeight = getItemWeight(item);
+				// And if quantity is specified, measure for specified quantity rather than whole stack
+				Long itemWeight = (requestQuantity != null) ? requestQuantity*itemSingleWeight : getItemWeight(item);
+
+				// Again, not sure if this check is necessary, but leaving it for now
 				if (itemWeight==null) itemWeight = 0L;
+
 				// If the item has a maxWeight, we will treat it as a container and include it's contents in the weight calculation..
 				if (item.getProperty("maxWeight")!=null)
 				{
 					Long itemCarryingWeight = getItemCarryingWeight(item);
 					itemWeight+=itemCarryingWeight;
 				}
-				
-				if (itemWeight>0L)
+
+				Long characterAvailableWeight = (Long)getCharacterMaxCarryingWeight(character) - getCharacterCarryingWeight(character);
+
+				// If the item is weightless, we skip the following check
+				stackTooHeavy = ((characterAvailableWeight < itemWeight) && (itemSingleWeight>0L));
+
+				if (stackTooHeavy)
 				{
-					long carrying = getCharacterCarryingWeight(character);
-					long maxCarrying = getCharacterMaxCarryingWeight(character);
-					
-					if (carrying+itemWeight>maxCarrying)
-						throw new UserErrorMessage("You cannot carry any more stuff! You are currently carrying "+GameUtils.formatNumber(carrying)+" grams and can carry a maximum of "+GameUtils.formatNumber(maxCarrying)+" grams.");
+					requestQuantity = characterAvailableWeight / itemSingleWeight;
+					if (requestQuantity==0L)
+						throw new UserErrorMessage("You try to pick it up, but it's too heavy! You can only carry "+GameUtils.formatNumber(characterAvailableWeight)+" grams more.");
 				}
-				
 			}
 		}
 		
-		
-		
 		if (handled==false)
 			throw new UserErrorMessage("Unable to move this item. It is probably no longer there. Try hitting the refresh button at the top of this popup?");
 
+		// With earlier checks, requestQuantity must be either null or greater than zero
+		// If not null, then either the user request passed container checks or request was lowered meet container capacity
+		if (requestQuantity != null && requestQuantity < itemQuantity)
+		{
+			final Key itemKey = item.getKey();
+			final String itemKind = item.getKind();
+			final Long oldQuantity = itemQuantity;
+			final Long newQuantity = requestQuantity;
+
+			try {
+				item = (CachedEntity) new Transaction<CachedEntity>(ds) {
+
+					@Override
+					public CachedEntity doTransaction(CachedDatastoreService ds) throws AbortTransactionException {
+
+						CachedEntity item = getEntity(itemKey);
```

Comment:
I think there is a refetch() method which you might want to use instead of getEntity(). It's specifically made for refetching entities that were originally fetched outside of a transaction for use inside of a transaction.